backend/src/layer/logic/image_review/image_reviewer.rs

```diff
@@ -0,0 +1,239 @@
+use async_trait::async_trait;
+use chrono::Local;
+use thiserror::Error;
+use tracing::log::warn;
+
+use crate::{
+    interface::{
+        image_hoster::ImageHoster,
+        image_review::ImageReviewScheduling,
+        persistent_data::{model::Image, ImageReviewDataAccess, Result},
+    },
+    util::Uuid,
+};
+
+pub type ReviewerResult<T> = std::result::Result<T, ReviewerError>;
+
+/// Enum describing the possible ways, the mail notification can fail.
+#[derive(Debug, Error, PartialEq, Eq)]
+pub enum ReviewerError {
+    #[error("an error occurred while getting the images")]
+    ImageGetError,
+    #[error("an error occurred while deleting the non-existent image {0:#?}")]
+    DeleteError(Uuid),
+    #[error("an error occurred while checking the image with id {0:#?} for its existence")]
+    CheckError(Uuid),
+    #[error("an error occurred while marking the image with id {0:#?} as checked")]
+    MarkError(Uuid),
+}
+
+const NUMBER_OF_IMAGES_TO_CHECK: u32 = 500;
+
+pub struct ImageReviewer<D, H>
+where
+    D: ImageReviewDataAccess,
+    H: ImageHoster,
+{
+    data_access: D,
+    image_hoster: H,
+}
+
+#[async_trait]
+impl<D, H> ImageReviewScheduling for ImageReviewer<D, H>
+where
+    D: ImageReviewDataAccess,
+    H: ImageHoster,
+{
+    async fn start_image_review(&self) {
+        let today = Local::now().date_naive();
+        self.review_images(
+            self.data_access
+                .get_n_images_by_rank_date(NUMBER_OF_IMAGES_TO_CHECK, today)
+                .await,
+        )
+        .await;
+        self.review_images(
+            self.data_access
+                .get_n_images_next_week_by_rank_not_checked_last_week(NUMBER_OF_IMAGES_TO_CHECK)
+                .await,
+        )
+        .await;
+        self.review_images(
+            self.data_access
+                .get_n_images_by_last_checked_not_checked_last_week(NUMBER_OF_IMAGES_TO_CHECK)
+                .await,
+        )
+        .await;
+    }
+}
+
+impl<D, H> ImageReviewer<D, H>
+where
+    D: ImageReviewDataAccess,
+    H: ImageHoster,
+{
+    #[must_use]
+    pub const fn new(data_access: D, image_hoster: H) -> Self {
+        Self {
+            data_access,
+            image_hoster,
+        }
+    }
+
+    async fn review_images(&self, images: Result<Vec<Image>>) {
+        if let Err(error) = self.try_review_images(images).await {
+            warn!("{error:#?}");
+        }
+    }
+
+    async fn try_review_images(&self, images: Result<Vec<Image>>) -> ReviewerResult<()> {
+        let images = images.map_err(|_e| ReviewerError::ImageGetError)?;
```

Comment:
we dont want to discard the data error as we lose context this way. We shoud rather inject the error in the ReviewerError like we did at other places (then you can also use the questionmark operator).
Also, passing the result as an argument makes me not so happy. Inserting the error just to get it back out seems odd to me.
Maybe it just would be better to do the database requests from inside the try_review_images method where you can return errors using `?`.

Reply:
we should also consider running these image reviews in parallel as 1500 of them could take some (not negligable) amout of time. This can be done with `join_all` over a iterator of futures from the `futures` crate (which is already used in swka_html_request) https://docs.rs/futures/latest/futures/future/fn.join_all.html